design-pattern.md

```diff
@@ -429,3 +429,287 @@ public class Main {
+
+#### Esempio
+
+Supponiamo di voler costruire un sistema di gestione di dati di un'azienda. L'azienda ha un database che contiene i dati dei dipendenti, e un sistema che permette di visualizzare questi dati. Il sistema di visualizzazione deve essere aggiornato ogni volta che il database cambia. Usiamo un **Observer**.
+
+Per prima cosa, definiamo l'interfaccia `Observer`, in questo esempio la chiamiamo `Visualizzazione`, che definisce il metodo `aggiorna()`:
+
+```java
+public interface Visualizzazione {
+  public void aggiorna();
+}
+```
+
+Definiamo le varie classi `ConcreteObserver`, in questo esempio le chiamiamo `VisualizzazioneConsole`, `VisualizzazioneGUI`, ecc., che implementano `Visualizzazione`:
+
+```java
+public class VisualizzazioneConsole 
+implements Visualizzazione {
+  @Override
+  public void aggiorna() {
+    System.out.println(
+      "Aggiornamento della 
+      visualizzazione console");
+  }
+}
+```
+
+```java
+public class VisualizzazioneGUI 
+implements Visualizzazione {
+  @Override
+  public void aggiorna() {
+    System.out.println(
+      "Aggiornamento della 
+      visualizzazione GUI");
+  }
+}
+```
+
+Definiamo la classe `Subject`, in questo esempio la chiamiamo `Database`, che ha un `ArrayList` di `Visualizzazione` e un metodo `aggiornaVisualizzazioni()` che chiama il metodo `aggiorna()` di ogni `Visualizzazione`:
+
+```java
+public class Database {
+  private ArrayList<Visualizzazione> 
+    visualizzazioni;
+
+  public Database() {
+    visualizzazioni = new ArrayList<>();
+  }
+
+  public void aggiungiVisualizzazione
+  (Visualizzazione visualizzazione) {
+    visualizzazioni.add(visualizzazione);
+  }
+
+  public void rimuoviVisualizzazione
+  (Visualizzazione visualizzazione) {
+    visualizzazioni.remove(visualizzazione);
+  }
+
+  protected void aggiornaVisualizzazioni() {
+    for(Visualizzazione visualizzazione
+    : visualizzazioni) {
+      visualizzazione.aggiorna();
+    }
+  }
+}
+```
+
+Si noti che nei metodi per aggiungere o rimuovere una `Visualizzazione`, abbiamo l'interfaccia `Visualizzazione` come tipo di parametro. Questo ci permette di aggiungere dinamicamente qualsiasi tipo di `Visualizzazione` che implementa l'interfaccia `Visualizzazione`.
+
+---
+
+<!-- _class: due -->
+
+### State
+
+**State** è un pattern che permette di cambiare il comportamento di un oggetto in base allo stato in cui si trova. È molto simile a **Strategy**.
+
+#### UML
+
+![](./immagini/state.svg)
+
+In questo diagramma, troviamo `State` che è l'interfaccia che definisce uno stato _generico_, `ConcreteState` che estende `State` e definisce il comportamento in quello specifico stato e `Context` che è l'oggetto che cambia stato.
+
+#### Esempio
```

Comment:
La descrizione è perfetta!
Però, anche se la differenza è sottile, nell'esempio hai usato Strategy e non State.
Se vedi l'esempio nelle slide, nella classe astratta `TCPState` oltre ai metodi `open`, `close`, e `acknowledge` che sono equivalenti al tuo metodo `comportamento`, c'è un metodo `transition` che caratterizza la differenza tra Strategy e State: in State, quando invochi il comportamento, può avvenire un cambiamento dello Stato che cambierà l'effetto della successiva chiamata a comportamento.
Infatti poi nello stato concreto `TCPListen`, durante la `open` c'è una traniszione di stato a `TCPEstablished` che è trasparente rispetto alla classe `TCPConnection` che semplicemente richiama  `open` sullo stato corrente.
Nel tuo caso invece la transizione di stato la deve fare l'utente con il metodo `setStato`, ma questo in realtà è proprio uno Strategy dove è l'utente a cambiare il comportamento della classe a runtime.
Btw per oggi mi sa che ho finito il tempo per le review 😆 , devo mettermi a fare TDE che sto indietrissimo 😭 .

Reply:
Aspè la PR è diventata una draft, forse non avevi ancora finito, in caso mi scuso. Vado a studiare che è meglio 😆

Reply:
ok, ho fatto lo stesso esempio delle slide per state e command